File: backend/models/UserProgress.js
// models/UserProgress.js
const mongoose = require('mongoose');

const UserProgressSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  design_pattern_id: { type: mongoose.Schema.Types.ObjectId, ref: 'DesignPattern', required: true },
  learning_completed: { type: Boolean, default: false },
  practice_completed: { type: Number, default: 0 }, // Number of correct answers
  test_completed: { type: Boolean, default: false },
  points: { type: Number, default: 0 },
  progress: { type: Number, default: 0 }, // Progress percentage
  created_at: { type: Date, default: Date.now },
  updated_at: { type: Date, default: Date.now },
});

module.exports = mongoose.model('UserProgress', UserProgressSchema);

File: dpla/lib/models/user_progress.dart
// lib/models/user_progress.dart

import 'package:json_annotation/json_annotation.dart';

part 'user_progress.g.dart';

@JsonSerializable()
class UserProgress {
  final String designPattern;
  final int progress; // Progress percentage
  final int points;
  final bool learningCompleted;
  final int practiceCompleted;
  final bool testCompleted;
  final DateTime updatedAt;

  UserProgress({
    required this.designPattern,
    required this.progress,
    required this.points,
    required this.learningCompleted,
    required this.practiceCompleted,
    required this.testCompleted,
    required this.updatedAt,
  });

  factory UserProgress.fromJson(Map<String, dynamic> json) =>
      _$UserProgressFromJson(json);
  Map<String, dynamic> toJson() => _$UserProgressToJson(this);
}
